// swagger-md/dist/generators/request-parameters-table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createParametersTable;

var _typePicker = require('./type-picker');

var _typePicker2 = _interopRequireDefault(_typePicker);

var _mdTable = require('./md-table');

var _mdTable2 = _interopRequireDefault(_mdTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createParametersTable(params) {
  if (!params.length) {
    return undefined;
  }

  return (0, _mdTable2.default)(['in', 'name', 'type', 'required', 'description', 'pattern', 'range', 'default', 'unique', 'multiple_of'], params.map(param => ({
    in: param.in,
    name: param.name,
    type: formatParamType(param),
    required: !!param.required,
    description: param.description,
    pattern: formatCode(param.pattern, false),
    range: formatParamRange(param),
    default: formatCode(param.default, true),
    unique: param.uniqueItems,
    multiple_of: param.multipleOf
  })));
}

function formatParamType(param) {
  if (param.$ref) {
    return _typePicker2.default.formatRef(param);
  }
  if (param.schema) {
    if (param.schema.$ref) {
      return _typePicker2.default.formatRef(param.schema);
    }
    return '';
  }

  let type = param.type;

  if (!type) {
    return '';
  }
  if (type === 'array') {
    type = `${type}, ${param.collectionFormat || 'csv'}`;
    if (param.items) {
      type = `${type} of ${formatParamType(param.items)}`;
    }
  }
  const number_type = _typePicker2.default.getNumberType(param);
  if (number_type) {
    return number_type;
  }

  const format = param.format;

  if (format) {
    return `${type}, ${format}`;
  }
  if (type === 'string' && param.enum) {
    return `${type}: ${param.enum.join(', ')}`;
  }
  return type;
}

function formatParamRange(param) {
  return formatRange(param, 'number', param.minimum, param.maximum, param.exclusiveMinimum, param.exclusiveMaximum) || formatRange(param, 'length', param.minLength, param.maxLength, false, false) || formatRange(param, 'items', param.minItems, param.maxItems, false, false);
}

function formatRange(param, type, min, max, exclusive_min, exclusive_max) {
  const has_min = typeof min === 'number';
  const has_max = typeof max === 'number';
  if (!has_min && !has_max) {
    return undefined;
  }
  let r = '';
  if (has_min) {
    r = `${min} ${exclusive_min ? '<' : '<='} `;
  }
  if (type === 'number') {
    r += _typePicker2.default.getNumberType(param) || type;
  } else {
    r += type;
  }
  if (has_max) {
    r += ` ${exclusive_max ? '<' : '<='} ${max}`;
  }
  return `\`${r}\``;
}

function formatCode(value, as_json) {
  if (value === undefined) {
    return undefined;
  }
  if (as_json) {
    return `\`${JSON.stringify(value)}\``;
  }
  return `\`${value}\``;
}
